dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/validator/relationship/DuplicationValidator.java

```diff
@@ -27,62 +27,36 @@
  */
 package org.hisp.dhis.tracker.validation.validator.relationship;
 
-import static org.hisp.dhis.tracker.TrackerType.ENROLLMENT;
-import static org.hisp.dhis.tracker.TrackerType.EVENT;
-import static org.hisp.dhis.tracker.TrackerType.TRACKED_ENTITY;
-import static org.hisp.dhis.tracker.validation.ValidationCode.E4012;
+import static org.hisp.dhis.tracker.validation.ValidationCode.E4018;
 import static org.hisp.dhis.tracker.validation.validator.relationship.ValidationUtils.getUidFromRelationshipItem;
 import static org.hisp.dhis.tracker.validation.validator.relationship.ValidationUtils.relationshipItemValueType;
 
-import java.util.Optional;
-
-import org.hisp.dhis.tracker.TrackerType;
+import org.hisp.dhis.tracker.TrackerImportStrategy;
 import org.hisp.dhis.tracker.bundle.TrackerBundle;
 import org.hisp.dhis.tracker.domain.Relationship;
-import org.hisp.dhis.tracker.domain.RelationshipItem;
 import org.hisp.dhis.tracker.validation.Reporter;
 import org.hisp.dhis.tracker.validation.Validator;
-import org.hisp.dhis.tracker.validation.validator.ValidationUtils;
 
-/**
- * @author Morten Svanæs <msvanaes@dhis2.org>
- */
-class DataRelationsValidator
-    implements Validator<Relationship>
+public class DuplicationValidator implements Validator<Relationship>
```

Comment:
@enricocolasante I just noticed we have https://github.com/dhis2/dhis2-core/blob/master/dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preprocess/DuplicateRelationshipsPreProcessor.java

which runs before validation. So is there even any work for this Validator i.e. can we remove it?

Reply:
The validator and the preprocessor are doing 2 slightly different things.
The preprocessor is removing duplicated relationships from the payload.
The validator is making sure that we are not adding to the DB a relationship that is already there